### PR TITLE
Python-kivy: init at 1.10.0

### PR DESCRIPTION
**Pull request by vanschelven**
###### Motivation for this change

Kivy is an Open source Python library for rapid development of applications
that make use of innovative user interfaces, such as multi-touch apps.

Most of Kivy's dependencies are optional, so defining a good "default" set of useful dependencies was a bit of a challenge. The present set of dependencies works well for me, but more dependencies, or even options for configuring the package, may be required in the future.

There is also some discussion at https://github.com/kivy/kivy/issues/5367

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] Linux
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: pkgs/development/python-modules/kivy/default.nix [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+assert stdenv.lib.versionAtLeast self.cython.version "0.23";
+
+# Rather than depending on a specific older version of Cython, we accept the
+# off-chance of breakage:
+# assert stdenv.lib.versionOlder self.cython.version "0.25.3";
+
+buildPythonPackage rec {
+  name = "Kivy-1.10.0";
+
+  src = pkgs.fetchurl {
```
Comment (FRidh) [inline pkgs/development/python-modules/kivy/default.nix]: fetchPypi

File: pkgs/development/python-modules/kivy/default.nix [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+    # Not done yet:
+    # "Camera is also unclear in the cross-platform sense, though for a
+    # canonical linux configuration the answer is probably opencv. SDL2 for
+    # image provider qualifies as "canonical", though it is my impression that
+    # ffpyplayer has wider and better format support."
+    ];
+
+  propagatedBuildInputs = with self; [
+    pillow
+    pkgs.mtdev
```
Comment (FRidh) [inline pkgs/development/python-modules/kivy/default.nix]: Why is this library propagated?

File: pkgs/top-level/python-packages.nix [hunk 3]
```diff
@@ -11176,6 +11176,29 @@ in {
 
   keyring = callPackage ../development/python-modules/keyring { };
 
+  kivy = callPackage ../development/python-modules/kivy { inherit stdenv buildPythonPackage maintainers platforms licenses pkgs self ; };
+
+  kivygarden = buildPythonPackage rec {
```
Comment (FRidh) [inline pkgs/top-level/python-packages.nix]: in separate file

File: pkgs/development/python-modules/kivy/default.nix [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+# tested there is no guarantee."
+
+assert stdenv.lib.versionAtLeast self.cython.version "0.23";
+
+# Rather than depending on a specific older version of Cython, we accept the
+# off-chance of breakage:
+# assert stdenv.lib.versionOlder self.cython.version "0.25.3";
+
+buildPythonPackage rec {
+  name = "Kivy-1.10.0";
```
Comment (FRidh) [inline pkgs/development/python-modules/kivy/default.nix]: can you include `pname`, `version`, and `${pname}-${version}` as is now more commonly done.

File: pkgs/top-level/python-packages.nix [hunk 5]
```diff
@@ -11176,6 +11176,29 @@ in {
...
+    # The sole reason for this package is that it is a (questionable) dependency of Kivy
+    name = "kivy-garden-0.1.4";
+
+    src = pkgs.fetchurl {
```
Comment (FRidh) [inline pkgs/top-level/python-packages.nix]: fetchPypi

File: pkgs/top-level/python-packages.nix [hunk 6]
```diff
@@ -11176,6 +11176,29 @@ in {
 
   keyring = callPackage ../development/python-modules/keyring { };
 
+  kivy = callPackage ../development/python-modules/kivy { inherit stdenv buildPythonPackage maintainers platforms licenses pkgs self ; };
+
+  kivygarden = buildPythonPackage rec {
+    # The sole reason for this package is that it is a (questionable) dependency of Kivy
+    name = "kivy-garden-0.1.4";
```
Comment (FRidh) [inline pkgs/top-level/python-packages.nix]: same with pname and version

File: pkgs/top-level/python-packages.nix [hunk 7]
```diff
@@ -11176,6 +11176,29 @@ in {
...
+
+    src = pkgs.fetchurl {
+      url = "mirror://pypi/k/kivy-garden/${name}.tar.gz";
+      sha256 = "0wkcpr2zc1q5jb0bi7v2dgc0vs5h1y7j42mviyh764j2i0kz8mn2";
+    };
+
+    buildInputs = with self; [ requests ];
```
Comment (FRidh) [inline pkgs/top-level/python-packages.nix]: Is this not needed during runtime? Or is it needed for the tests? In case of the latter, use `checkInputs`.

**Review by FRidh — CHANGES_REQUESTED** [review]

File: pkgs/development/python-modules/kivy/default.nix (hunk 2 above)
Comment (vanschelven) [reply pkgs/development/python-modules/kivy/default.nix]: This is a mistake; I thought that would be helpful, but the actual fix was found later (https://groups.google.com/forum/#!topic/nix-devel/CPQKpM3u1EY), and I never downgraded to non-propagated. Will fix.

File: pkgs/top-level/python-packages.nix (hunk 7 above)
Comment (vanschelven) [reply pkgs/top-level/python-packages.nix]: It's needed during runtime.

kivy-garden is a "not invented here" package manager for kivy components, and as such uses `requests` to fetch those components. 

The real problem is that such a thing exists in the first place, and that it is an explicit dependency of Kivy. I'll talk with the kivy people about both of those issues.

In the meantime, given the fact that the present PR makes kivy-garden available, it should do so in a non-broken way, so I'll promote this to a `propagatedBuildInput`

File: pkgs/development/python-modules/kivy/default.nix [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+  # can't retrieve git version. We remove the git-invocation from setup.py
+  # This patch can presumably be removed on the next version of Kivy, because the problem
+  # was addressed in https://github.com/kivy/kivy/issues/5365
+  patches = [
+    ./setup.py.patch
+  ];
+
+  buildInputs = with self; [
+    pkgconfig
+    cython
```
Comment (FRidh) [inline pkgs/development/python-modules/kivy/default.nix]: these two can be `nativeBuildInputs` if I am correct.

File: pkgs/development/python-modules/kivy/default.nix [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+
+    kivygarden  # See https://github.com/kivy/kivy/issues/5367 suggestion I-2
+    requests  # See https://github.com/kivy/kivy/issues/5367 suggestion I-3
+    docutils # See https://github.com/kivy/kivy/issues/5367 question Q-5
+    pygments # See https://github.com/kivy/kivy/issues/5367 question Q-6
```
Comment (FRidh) [inline pkgs/development/python-modules/kivy/default.nix]: Are these 4 only needed during build time? Or during runtime? If the latter -> `propagatedBuildInputs`

File: pkgs/development/python-modules/kivy/default.nix [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+    # The setup below is, as much as I can figure out how to make it, an attempt
+    # to create a "canonical full install" of Kivy on linux, as allued to in
+    # https://github.com/kivy/kivy/issues/5367 question Q-7
+
+    # Some sources to reconstruct such a thing are:
+    # .travis.yml (which contains the actual build instructions for CI)
+    # https://kivy.org/docs/installation/installation.html
+    # https://kivy.org/docs/installation/installation-linux.html
+
+    pkgs.mesa
```
Comment (FRidh) [inline pkgs/development/python-modules/kivy/default.nix]: get rid of `pkgs`

File: pkgs/development/python-modules/kivy/kivy-garden.nix [hunk 11]
```diff
@@ -0,0 +1,25 @@
+# The sole reason for this package is that it is a (questionable) dependency of Kivy
+{ buildPythonPackage, fetchPypi, maintainers, platforms, licenses, pkgs, self }:
```
Comment (FRidh) [inline pkgs/development/python-modules/kivy/kivy-garden.nix]: same as the other

File: pkgs/development/python-modules/kivy/kivy-garden.nix [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+  buildInputs = with self; [ ];
+  propagatedBuildInputs = with self; [ requests ];
+
+  meta = {
+    description = "The kivy garden installation script, split into its own package for convenient use in buildozer.";
+
+    homepage    = "https://pypi.python.org/pypi/kivy-garden";
+    license     = licenses.mit;
+    maintainers = with maintainers; [ vanschelven ];
+    platforms   = platforms.unix;  # Can only test linux; in principle other platforms are supported
```
Comment (FRidh) [inline pkgs/development/python-modules/kivy/kivy-garden.nix]: `buildPythonPackage` already sets platforms

File: pkgs/development/python-modules/kivy/default.nix [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+   substituteInPlace kivy/lib/mtdev.py \
+     --replace "LoadLibrary('libmtdev.so.1')" "LoadLibrary('${pkgs.mtdev}/lib/libmtdev.so.1')"
+  '';
+
+  meta = {
+    description = "A software library for rapid development of hardware-accelerated multitouch applications.";
+    homepage    = "https://pypi.python.org/pypi/kivy";
+    license     = licenses.mit;
+    maintainers = with maintainers; [ vanschelven ];
+    platforms   = platforms.unix;  # I can only test linux; in principle other platforms are supported
```
Comment (FRidh) [inline pkgs/development/python-modules/kivy/default.nix]: `buildPythonPackage` already sets platforms

File: pkgs/top-level/python-packages.nix [hunk 14]
```diff
@@ -11176,6 +11176,10 @@ in {
 
   keyring = callPackage ../development/python-modules/keyring { };
 
+  kivy = callPackage ../development/python-modules/kivy { inherit stdenv buildPythonPackage fetchPypi maintainers platforms licenses pkgs self ; };
```
Comment (FRidh) [inline pkgs/top-level/python-packages.nix]: callPackage will populate the arguments automatically

**Review by FRidh — CHANGES_REQUESTED** [review]

File: pkgs/development/python-modules/kivy/default.nix (hunk 8 above)
Comment (vanschelven) [reply pkgs/development/python-modules/kivy/default.nix]: I am not confident that I understand the difference (see https://groups.google.com/forum/#!topic/nix-devel/Ad1b0RShe6k) and all other Python packages that I could find do in fact use `buildInputs` here, so I'm a bit wary to proceed with this tip. Feel free to correct me if you're sure about this behavior though.

File: pkgs/development/python-modules/kivy/default.nix [hunk 15]
```diff
@@ -0,0 +1,101 @@
+{ stdenv, buildPythonPackage, fetchPypi, maintainers, platforms, licenses, pkgs, self }:
```
[the remark this answers is not in the document]
Comment (vanschelven) [reply pkgs/development/python-modules/kivy/default.nix]: I'm having some trouble implementing this advice, and associated advice on how to remove the associated explicit parameter-setting on the calling side. Could you show me how it's done?

File: pkgs/development/python-modules/kivy/default.nix (hunk 9 above)
Comment (vanschelven) [reply pkgs/development/python-modules/kivy/default.nix]: Regarding docutils: purely build time.

Regarding the rest: `setup.py` refuses to install without these packages present, but they are fundamentally optional and (in the case of kivygarden especially) it's undesirable to pollute the runtime environment too much with their presence. If someone e.g. want the `pygments` functionality at runtime, it's better to have them install this explicitly than to "pull in the world"

**Comment by vanschelven** [comment]
@FRidh I've responded in the above remarking that I'm not confident enough in my knowledge of Nix packaging to push forwards from here. Any chance you could implement your own remarks/requests so that we can push this PR forward and get it accepted?

**Comment by vanschelven** [comment]
I'm closing this pull request; it no longer works against the current version of master, and updating is non-trivial at the moment. I might reopen a similar request at a later point in time.

**Pull request closed without merging**